scripts/convert_ipynb_to_mdx.py

```diff
@@ -0,0 +1,1080 @@
+        plot_data_folder (Path): Path to the folder where plot data should be
+            stored.
+
+    Returns:
+        str: MDX formatted code cell.
+    """
+    cell_input_mdx = handle_cell_input(cell, "python")
+    cell_output_mdx = handle_cell_outputs(cell, plot_data_folder)
+    return cell_input_mdx + cell_output_mdx
+
+
+def transform_notebook(path: Path, nb_metadata: object) -> str:
+    """
+    Transform a notebook located at the given path into MDX.
+
+    Args:
+        path (Path): Path to the Jupyter notebook tutorial.
+
+    Returns:
+        str: MDX formatted string.
+    """
+    filename, assets_folder = create_folders(path)
+    img_folder = assets_folder / "img"
+    plot_data_folder = assets_folder / "plot_data"
+    save_folder = assets_folder.joinpath("..").resolve()
+    nb = load_notebook(path)
+    mdx = ""
+    mdx += create_frontmatter(path, nb_metadata)
+    mdx += create_imports()
+    mdx += create_buttons(nb_metadata)
+    for cell in nb["cells"]:
+        cell_type = cell["cell_type"]
+
+        # Handle a Markdown cell.
+        if cell_type == "markdown":
+            mdx += handle_markdown_cell(cell, img_folder, LIB_DIR)
+
+        # Handle a code cell.
+        if cell_type == "code":
+            mdx += handle_code_cell(cell, plot_data_folder)
+
+    # Write the MDX file to disk.
+    save_path = save_folder / "index.mdx"
+    with save_path.open("w") as f:
+        f.write(mdx)
+
+    # Return the string for debugging purposes.
+    return mdx
+
+
+def clean_up_directories() -> None:
+    """
+    Delete output from previously converted notebooks. Particularly useful for
+    removing plot data since those filenames are randomly generated and won't
+    be replaced with future runs.
+
+    Returns:
+        None: Does not return anything.
+    """
+    if TUTORIALS_DIR.exists():
+        # We intentionally leave the static `index.mdx` file in place since that is not
+        # autogenerated.
+        for item in os.scandir(TUTORIALS_DIR):
+            if item.is_dir():
+                shutil.rmtree(item.path)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Convert tutorial notebooks into mdx files."
+    )
+    parser.add_argument(
+        "--clean",
+        action="store_true",
+        help="Delete output from previously converted notebooks.",
+    )
+    args = parser.parse_args()
+
+    tutorials_metadata = load_nb_metadata()
+    print("--------------------------------------------")
+    print("Converting tutorial notebooks into mdx files")
+    print("--------------------------------------------")
+    if args.clean:
+        clean_up_directories()
+    for metadata in tutorials_metadata:
+        path = (
+            LIB_DIR / "tutorials" / metadata["id"] / (metadata["id"] + ".ipynb")
+        ).resolve()
+        print(f"{path.stem}")
+        mdx = transform_notebook(path, metadata)
+    print("")
```

Comment:
I guess we still need 1000 lines of custom conversion code for the notebooks :( We can't use a third party tool like nbconvert to avoid the need for maintaining our own scripts for this?

Reply:
Docusaurus only natively accepts MD and MDX for docs. There's a hacky way to force html into the page and that's what we were previously doing but it's much less flexible and doesn't get picked up by docusaurus's sidebar or versioning system.

There _is_ an MD output option in nbconvert but we would lose the interactive plots and html/react features of MDX. It might be possible to shave down some of this script by pre-processing using nbconvert but I think the loss of ipynb metadata might actually make some parts more difficult to parse/convert.

This file is mostly a copy of a script written by the BeanMachine team so there may be pieces of it that are not relevant to our tutorials and can trim. Here's the original: https://github.com/facebookresearch/beanmachine/blob/main/website/scripts/convert_ipynb_to_mdx.py

And here's the PR where I originally added support for it in the Ax repo first with the most context: https://github.com/CristianLara/Ax/pull/13

Reply:
This is definitely the most complex part of the new website and I also would prefer to avoid it if there is a better alternative.

Reply:
> This file is mostly a copy of a script written by the BeanMachine team so there may be pieces of it that are not relevant to our tutorials and can trim

Yeah at the very least we'd want to cut this down to only the parts that are essential to our setup.